### PR TITLE
FOGL-7303 System tests for the newly introduced linked data type support in the OMF plugin.

### PR DESCRIPTION
**Pull request by Mohit04tomar**
http://jenkins.dianomic.com:8080/job/fledge_package_system_test_pi_web_api_2021sp3_linked_data_type/2/

### CODE REVIEW

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+        
+        data = {"readingsPerSec": "{}".format(poll_rate)}
+        put_url="/fledge/category/{0}_{1}Advanced".format(ASSET, south_plugin)
+        utils.put_request(fledge_url, urllib.parse.quote(put_url, safe='?,=,&,/'), data)
+        
+        enable_schedule(fledge_url, "{0}_{1}".format(ASSET, south_plugin))
+        poll_rate+=5
+    
+    start_north_omf_as_a_service(fledge_url, pi_host, pi_port, pi_user=pi_admin, pi_pwd=pi_passwd, pi_use_legacy="false",
+                                 service_name=NORTH_INSTANCE_NAME, default_af_location=AF_HIERARCHY_LEVEL)
```
Comment (YashTatkondawar) [inline tests/system/python/packages/test_pi_webapi_linked_data_type.py:217]: Check indentation

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+
+class Test_linked_data_PIWebAPI:
+    def test_linked_data(self, clean_setup_fledge_packages, reset_fledge, start_south_north, fledge_url, 
+                         pi_host, pi_admin, pi_passwd, pi_db, wait_time, retries, pi_port, disable_schedule, 
+                         verify_hierarchy_and_get_datapoints_from_pi_web_api, clear_pi_system_through_pi_web_api, 
+                         skip_verify_north_interface, asset_name=ASSET):
```
Comment (YashTatkondawar) [inline tests/system/python/packages/test_pi_webapi_linked_data_type.py:223]: Check indentation of lines

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+    for plugin in SOUTH_PLUGINS_LIST:
+        filter_cfg = {"enable": "true", "expression": "log({}_{}-Ingest)".format(ASSET, plugin)}
+        add_filter("expression", None, "{}".format(plugin), filter_cfg, fledge_url, "{}_{}".format(ASSET, plugin), installation_type='package')
+
+@pytest.fixture
+def start_south_north(add_south, start_north_task_omf_web_api, add_filter, remove_data_file,
+                      fledge_url, pi_host, pi_port, pi_admin, pi_passwd, pi_db,
+                      start_north_omf_as_a_service, enable_schedule, 
+                      clear_pi_system_through_pi_web_api, asset_name=ASSET):
+    """ This fixture starts the sinusoid plugin and north pi web api plugin. Also puts a filter
```
Comment (YashTatkondawar) [inline tests/system/python/packages/test_pi_webapi_linked_data_type.py:191]: Are we starting only sinusoid plugin?

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+        verify_statistics_map(fledge_url, skip_verify_north_interface)
+        verify_asset_tracking_details(fledge_url, skip_verify_north_interface)
+        
+        # Verify Data from fledge sent to PI Server is same.
+        verify_data_between_fledge_and_piwebapi(fledge_url, pi_host, pi_admin, pi_passwd, pi_db, AF_HIERARCHY_LEVEL, ASSET, SOUTH_PLUGINS_LIST, verify_hierarchy_and_get_datapoints_from_pi_web_api)
+        
+    def test_linked_data_reconfig(self, reset_fledge, start_south_north, fledge_url, add_configure_filter,
+                         pi_host, pi_admin, pi_passwd, pi_db, wait_time, retries, pi_port, disable_schedule, 
+                         verify_hierarchy_and_get_datapoints_from_pi_web_api, clear_pi_system_through_pi_web_api, 
+                         skip_verify_north_interface, asset_name=ASSET):
```
Comment (YashTatkondawar) [inline tests/system/python/packages/test_pi_webapi_linked_data_type.py:253]: Check indentation of lines

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+            read_data_from_pi: Fixture to read data from PI
+            skip_verify_north_interface: Flag for assertion of data using PI web API
+            Assertions:
+                on endpoint GET /fledge/ping
+                on endpoint GET /fledge/statistics
+                on endpoint GET /fledge/asset
+                on endpoint GET /fledge/asset/<asset_name>
+                data received from PI is same as data sent"""
+        
+        time.sleep(wait_time)
```
Comment (YashTatkondawar) [inline tests/system/python/packages/test_pi_webapi_linked_data_type.py:236]: Please add a comment why this sleep is required.

**Comment by YashTatkondawar** [comment]
In general check if all print statements are required or not.

File: tests/system/python/conftest.py [hunk 6]
```diff
@@ -423,6 +423,106 @@ def clear_pi_system_through_pi_web_api():
...
+    def _verify_hierarchy_and_get_datapoints_from_pi_web_api(host, admin, password, pi_database, af_hierarchy_list, asset, sensor):
+        """ This method verifies hierarchy created in pi web api is correctly """
+    
+        username_password = "{}:{}".format(admin, password)
+        username_password_b64 = base64.b64encode(username_password.encode('ascii')).decode("ascii")
+        headers = {'Authorization': 'Basic %s' % username_password_b64}
+        AF_HIERARCHY_LIST=af_hierarchy_list.split('/')[1:]
+        AF_HIERARCHY_COUNT=len(AF_HIERARCHY_LIST)
+        
+        print(AF_HIERARCHY_LIST)
```
Comment (deepanshu-yadav) [inline tests/system/python/conftest.py:437]: Print not required.

File: tests/system/python/conftest.py [hunk 7]
```diff
@@ -423,6 +423,106 @@ def clear_pi_system_through_pi_web_api():
...
+                else:
+                    raise Exception('Database not exist')
+                
+                if elements_url is not None:
+                    conn.request("GET", elements_url, headers=headers)
+                    res = conn.getresponse()
+                    r = json.loads(res.read().decode())
+                    items = r['Items']
+                    # pprint(items)
+                    print('-----------')
```
Comment (deepanshu-yadav) [inline tests/system/python/conftest.py:469]: Print not required.

**Review by deepanshu-yadav — CHANGES_REQUESTED** [review]

File: tests/system/python/conftest.py [hunk 8]
```diff
@@ -423,6 +423,104 @@ def clear_pi_system_through_pi_web_api():
...
+                                    count += 1
+                                    if item['Name'] in sensor:
+                                        record[item['Name']] = list(map(lambda val: val['Value'], item['Items']))
+                                        Item_matched = True
+                                    elif count == no_of_datapoint_in_pi_server and Item_matched == False:
+                                        raise "Required Data points is not Present --> {}".format(sensor)
+                        else:
+                            raise "Asset does not exist, Although Hierarchy is correct"
+                        print('no. of sinusoid DPs',len(record['sinusoid']))
+                        print('no. of randomwalk DPs', len(record['randomwalk']))
```
[the remark this answers is not in the document]
Comment (Mohit04tomar) [reply tests/system/python/conftest.py:510]: ok

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py [hunk 9]
```diff
@@ -0,0 +1,277 @@
...
+    for plugin in SOUTH_PLUGINS_LIST:
+        print("Before sort")
+        print('Data from fledge for {} --> '.format(plugin), data_from_fledge[plugin])
+        print('Data from PI for {} --> '.format(plugin), data_from_pi[plugin])
+        print("===============================================")
+        print("After Sort")
+        listA = sorted(data_from_fledge[plugin])
+        listB = sorted(data_from_pi[plugin])
+        print('Data from fledge for {} --> '.format(plugin), listA)
+        print('Data from PI for {} --> '.format(plugin), listB)
```
[the remark this answers is not in the document]
Comment (Mohit04tomar) [reply tests/system/python/packages/test_pi_webapi_linked_data_type.py]: Removed Unwanted print statements.

File: tests/system/python/packages/test_pi_webapi_linked_data_type.py (hunk 3 above)
Comment (Mohit04tomar) [reply tests/system/python/packages/test_pi_webapi_linked_data_type.py:182]: No, we are using siunsoid and randomwalk and starting both the plugins.

**Review by deepanshu-yadav — APPROVED** [review]

**Comment by Mohit04tomar** [comment]
This is the CI job build that ran after adding third test --> http://jenkins.dianomic.com:8080/view/PI%20stuff/job/fledge_package_system_test_pi_web_api_2021sp3_linked_data_type/3/

**Review by YashTatkondawar — APPROVED** [review]

**Review by deepanshu-yadav — APPROVED** [review]

**Pull request merged**